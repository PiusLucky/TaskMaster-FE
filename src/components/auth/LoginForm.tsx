"use client";

import React, { useState } from "react";
import * as z from "zod";
import { Eye, EyeOff } from "lucide-react";
import {
  Form,
  FormControl,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from "@/components/ui/form";

import { Input } from "@/components/ui/input";
import { useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import Link from "next/link";
import { apiClient, cookieStorageManager } from "@/api";
import { useSetAtom } from "jotai";
import { ATOMS } from "@/api/atoms";
import { storageKeys } from "@/api/storageKeys";
import { useRouter } from "next/navigation";
import apiResources from "@/api/resources";
import { ILoginData, IUserInfo } from "@/types/global-types";
import TMButton from "../common/TMButton";

const formSchema = z.object({
  email: z.string().email("Enter a valid email"),
  password: z
    .string()
    .min(5, {
      message: "Password must be at least 5 characters",
    })
    .max(50, {
      message: "password must be at most 50 characters",
    }),
});

function LoginForm() {
  const [showPassword, setShowPassword] = useState(false);
  const [loading, setLoading] = useState(false);
  const setError = useSetAtom(ATOMS.axiosError);
  const router = useRouter();

  const form = useForm<z.infer<typeof formSchema>>({
    resolver: zodResolver(formSchema),
    defaultValues: {
      email: "",
      password: "",
    },
  });

  async function onSubmit(values: z.infer<typeof formSchema>) {
    const { email, password } = values;
    setLoading(true);
    try {
      cookieStorageManager.clearAll();
      const { data } = await apiClient.post(
        apiResources.user,
        "/users/login",
        {
          email,
          password,
        },
        setError,
      );

      await afterLoginUpdate(data);
    } catch (err) {
      setLoading(false);
    }
  }

  const afterLoginUpdate = async (data: ILoginData) => {
    cookieStorageManager.addOrUpdateItem(storageKeys.token, data.access_token);

    const userInfo = await apiClient.get<IUserInfo>(
      apiResources.user,
      "/users",
      setError,
    );
    cookieStorageManager.addOrUpdateItem(storageKeys.user, userInfo.data);

    //Route to Dashboard
    router.push("/dashboard", { scroll: false });
  };

  const togglePaswordVisibility = () => {
    setShowPassword(!showPassword);
  };

  return (
    <div className="rounded-[2.5rem] shadow-lg p-[2.75rem] bg-white">
      <section className="mb-[1.81rem]">
        <div className="flex justify-between flex-col gap-2 md:flex-row">
          <div className="flex w-[200px] items-center gap-[0.63rem] -ml-2">
            <img
              src="/images/logo.png"
              alt="logo"
              className="w-[3rem] md:w-[4rem]"
            />
          </div>
          <div className="text-[#8D8D8D]">
            <p>No Account ?</p>
            <Link href="/auth/signup">
              <p className="text-primary  cursor-pointer font-bold inline-block">
                Sign up
              </p>
            </Link>
          </div>
        </div>
        <div className="text-[1.2rem] md:text-[1.5rem] font-medium mt-[0.81rem]">
          TaskMaster
        </div>
      </section>
      <section>
        <Form {...form}>
          <form onSubmit={form.handleSubmit(onSubmit)} className="space-y-6">
            {/* Email */}
            <FormField
              control={form.control}
              name="email"
              render={({ field }) => (
                <FormItem>
                  <FormLabel>Email</FormLabel>
                  <FormControl>
                    <div>
                      <Input
                        placeholder="e.g pauldoe@gmail.com"
                        {...field}
                        className="focus:!ring-primary h-[3.125rem]"
                        data-testid="email"
                      />
                    </div>
                  </FormControl>
                  <FormMessage />
                </FormItem>
              )}
            />

            {/* Password */}
            <FormField
              control={form.control}
              name="password"
              render={({ field }) => (
                <FormItem>
                  <FormLabel>Password</FormLabel>
                  <FormControl className="relative">
                    <div>
                      <Input
                        placeholder="Password"
                        {...field}
                        className="pr-[4.5rem] focus:!ring-primary h-[3.125rem]"
                        type={!showPassword ? "password" : "text"}
                        data-testid="password"
                      />
                      <div className="bg-[#D9D9D9] cursor-pointer absolute w-[3.6875rem] h-[2.375rem] rounded-[0.5rem] flex justify-center items-center font-bold top-[.35rem] right-1">
                        {showPassword ? (
                          <div onClick={() => togglePaswordVisibility()}>
                            <Eye />
                          </div>
                        ) : (
                          <div onClick={() => togglePaswordVisibility()}>
                            <EyeOff />
                          </div>
                        )}
                      </div>
                    </div>
                  </FormControl>
                  <FormMessage />
                </FormItem>
              )}
            />
            <div className="!mt-[3rem]">
              <TMButton
                text="Sign in"
                isSubmitable
                width="w-full"
                isLoading={loading}
                data-testid="signin-cta"
              />
            </div>
          </form>
        </Form>
      </section>
    </div>
  );
}


export default LoginForm;
