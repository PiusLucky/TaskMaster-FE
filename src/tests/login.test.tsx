import { expect, test } from "vitest";
import { render } from "@testing-library/react";
import LoginForm from "../components/auth/LoginForm";
import renderer from "react-test-renderer";

test("Email and password fields should render", async () => {
  const { getByTestId } = render(<LoginForm />);

  const email = getByTestId("email");
  const password = getByTestId("password");

  expect(email).to.exist;
  expect(password).to.exist;
});

test("The CTA must render", async () => {
  const { getByText } = render(<LoginForm />);
  const signinCTA = getByText("Sign in");
  expect(signinCTA).to.exist;
});

//Snapshot Testing
it("LoginForm should match snapshots", () => {
  const tree = renderer.create(<LoginForm />).toJSON();
  expect(tree).toMatchSnapshot();
});
