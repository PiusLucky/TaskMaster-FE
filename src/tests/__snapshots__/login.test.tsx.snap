// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm should match snapshots 1`] = `
<div
  className="rounded-[2.5rem] shadow-lg p-[2.75rem] bg-white"
>
  <section
    className="mb-[1.81rem]"
  >
    <div
      className="flex justify-between flex-col gap-2 md:flex-row"
    >
      <div
        className="flex w-[200px] items-center gap-[0.63rem] -ml-2"
      >
        <img
          alt="logo"
          className="w-[3rem] md:w-[4rem]"
          src="/images/logo.png"
        />
      </div>
      <div
        className="text-[#8D8D8D]"
      >
        <p>
          No Account ?
        </p>
        <a
          href="/auth/signup"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          <p
            className="text-primary  cursor-pointer font-bold inline-block"
          >
            Sign up
          </p>
        </a>
      </div>
    </div>
    <div
      className="text-[1.2rem] md:text-[1.5rem] font-medium mt-[0.81rem]"
    >
      TaskMaster
    </div>
  </section>
  <section>
    <form
      className="space-y-6"
      onSubmit={[Function]}
    >
      <div
        className="space-y-2"
      >
        <label
          className="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
          htmlFor=":r0:-form-item"
          onMouseDown={[Function]}
        >
          Email
        </label>
        <div
          aria-describedby=":r0:-form-item-description"
          aria-invalid={false}
          id=":r0:-form-item"
        >
          <input
            className="flex w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 focus:!ring-primary h-[3.125rem]"
            data-testid="email"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="e.g pauldoe@gmail.com"
            value=""
          />
        </div>
      </div>
      <div
        className="space-y-2"
      >
        <label
          className="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
          htmlFor=":r1:-form-item"
          onMouseDown={[Function]}
        >
          Password
        </label>
        <div
          aria-describedby=":r1:-form-item-description"
          aria-invalid={false}
          className="relative"
          id=":r1:-form-item"
        >
          <input
            className="flex w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 pr-[4.5rem] focus:!ring-primary h-[3.125rem]"
            data-testid="password"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
          <div
            className="bg-[#D9D9D9] cursor-pointer absolute w-[3.6875rem] h-[2.375rem] rounded-[0.5rem] flex justify-center items-center font-bold top-[.35rem] right-1"
          >
            <div
              onClick={[Function]}
            >
              <svg
                className="lucide lucide-eye-off"
                fill="none"
                height={24}
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth={2}
                viewBox="0 0 24 24"
                width={24}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9.88 9.88a3 3 0 1 0 4.24 4.24"
                />
                <path
                  d="M10.73 5.08A10.43 10.43 0 0 1 12 5c7 0 10 7 10 7a13.16 13.16 0 0 1-1.67 2.68"
                />
                <path
                  d="M6.61 6.61A13.526 13.526 0 0 0 2 12s3 7 10 7a9.74 9.74 0 0 0 5.39-1.61"
                />
                <line
                  x1="2"
                  x2="22"
                  y1="2"
                  y2="22"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        className="!mt-[3rem]"
      >
        <button
          className="inline-flex items-center justify-center text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 hover:bg-primary/90 h-10 px-4 py-2 bg-primary text-white shadow-xl w-full md:w-full md:h-[56px] select-none rounded-[12px] undefined"
          disabled={false}
          type="submit"
        >
          <div
            className="flex gap-2 items-center"
          >
            Sign in
          </div>
        </button>
      </div>
    </form>
  </section>
</div>
`;
